File: server/src/coherenceAnalyzer.ts
/**
 * Pure-JS Coherence Metric Parser (Module 3, Member 1).
 *
 * Fallback used when the dedicated Python NLP microservice is unreachable,
 * ensuring the Coherence Metric Parser functions reliably in any environment.
 * It measures causal language and cognitive insight/realization words,
 * computing density ratios and depth scores to track cognitive processing depth.
 */

export const CAUSAL_WORDS = new Set([
  "because", "since", "therefore", "thus", "consequently", "hence",
  "cause", "caused", "causes", "causing",
  "reason", "reasons", "why", "wherefore", "inasmuch",
  "result", "results", "resulted", "resulting",
  "effect", "effects", "affect", "affects", "affected", "affecting",
  "leads", "led", "leading",
  "depends", "depended", "depending",
  "outcome", "outcomes", "origin", "originated",
  "driven", "drives", "whereby",
]);

export const CAUSAL_PHRASES = [
  "as a result", "because of", "due to", "in order to", "led to",
  "leading to", "so that", "thanks to", "for this reason",
];

export const INSIGHT_WORDS = new Set([
  "realize", "realizes", "realized", "realizing", "realization",
  "understand", "understands", "understood", "understanding",
  "comprehend", "comprehends", "comprehended", "comprehending", "comprehension",
  "know", "knows", "knew", "knowing", "known", "knowledge",
  "think", "thinks", "thought", "thinking",
  "conclude", "concludes", "concluded", "concluding", "conclusion",
  "meaning", "meanings", "mean", "means", "meant",
  "insight", "insights", "insightful",
  "clarify", "clarifies", "clarified", "clarifying", "clarity",
  "figure", "figured", "figuring",
  "discern", "discerned", "discerning",
  "learn", "learns", "learned", "learning",
  "discover", "discovers", "discovered", "discovering", "discovery",
  "recognize", "recognizes", "recognized", "recognizing", "recognition",
  "perceive", "perceived", "perceiving", "perception",
  "ponder", "pondered", "pondering",
  "reflect", "reflected", "reflecting", "reflection",
  "acknowledge", "acknowledged", "acknowledging",
  "admit", "admitted", "admitting",
  "accept", "accepted", "accepting", "acceptance",
  "resolve", "resolved", "resolving", "resolution",
  "synthesize", "synthesized", "synthesizing", "synthesis",
]);

export interface DetectedWord {
  word: string;
  count: number;
}

export interface CoherenceAnalysis {
  totalWords: number;
  causalCount: number;
  causalDensity: number;
  insightCount: number;
  insightDensity: number;
  totalCoherenceCount: number;
  coherenceRatio: number;
  causalToInsightRatio: number | null;
  depthLevel: string;
  depthScore: number;
  detectedCausalWords: DetectedWord[];
  detectedInsightWords: DetectedWord[];
  tokenizer: string;
  source: "nlp" | "js-fallback";
}

const TOKEN_RE = /[a-zA-Z']+/g;

export function analyzeCoherence(text: string): CoherenceAnalysis {
  if (!text || !text.trim()) {
    return {
      totalWords: 0,
      causalCount: 0,
      causalDensity: 0,
      insightCount: 0,
      insightDensity: 0,
      totalCoherenceCount: 0,
      coherenceRatio: 0,
      causalToInsightRatio: null,
      depthLevel: "Raw / Descriptive Stance",
      depthScore: 0,
      detectedCausalWords: [],
      detectedInsightWords: [],
      tokenizer: "regex",
      source: "js-fallback",
    };
  }

  const lower = text.toLowerCase();
  const tokens: string[] = [];
  let m: RegExpExecArray | null;
  TOKEN_RE.lastIndex = 0;
  while ((m = TOKEN_RE.exec(lower)) !== null) {
    tokens.push(m[0]);
  }

  const totalWords = Math.max(tokens.length, 1);

  // 1. Check multi-word causal phrases
  const causalMap = new Map<string, number>();
  let phraseCausalCount = 0;
  for (const phrase of CAUSAL_PHRASES) {
    const escaped = phrase.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const re = new RegExp(`\\b${escaped}\\b`, "g");
    const matches = lower.match(re);
    if (matches && matches.length > 0) {
      causalMap.set(phrase, matches.length);
      phraseCausalCount += matches.length;
    }
  }

  // 2. Token-level counts
  const insightMap = new Map<string, number>();
  let tokenCausalCount = 0;
  let tokenInsightCount = 0;

  for (const token of tokens) {
    if (CAUSAL_WORDS.has(token)) {
      causalMap.set(token, (causalMap.get(token) ?? 0) + 1);
      tokenCausalCount++;
    } else if (INSIGHT_WORDS.has(token)) {
      insightMap.set(token, (insightMap.get(token) ?? 0) + 1);
      tokenInsightCount++;
    }
  }

  const totalCausal = phraseCausalCount + tokenCausalCount;
  const totalInsight = tokenInsightCount;
  const totalCoherence = totalCausal + totalInsight;

  const round = (n: number) => Math.round(n * 100) / 100;
  const causalDensity = round((totalCausal / totalWords) * 100);
  const insightDensity = round((totalInsight / totalWords) * 100);
  const coherenceRatio = round((totalCoherence / totalWords) * 100);

  let depthLevel = "Raw / Descriptive Stance";
  let depthScore = 0;
  if (coherenceRatio >= 6.0) {
    depthLevel = "High Cognitive Integration";
    depthScore = Math.min(100, Math.round(coherenceRatio * 12));
  } else if (coherenceRatio >= 3.0) {
    depthLevel = "Moderate Cognitive Processing";
    depthScore = Math.round(coherenceRatio * 12);
  } else if (coherenceRatio > 0) {
    depthLevel = "Emerging Causal Reflection";
    depthScore = Math.max(15, Math.round(coherenceRatio * 12));
  }

  const causalToInsightRatio =
    totalInsight > 0 ? round(totalCausal / totalInsight) : null;

  const detectedCausalWords: DetectedWord[] = Array.from(causalMap.entries())
    .map(([word, count]) => ({ word, count }))
    .sort((a, b) => b.count - a.count);

  const detectedInsightWords: DetectedWord[] = Array.from(insightMap.entries())
    .map(([word, count]) => ({ word, count }))
    .sort((a, b) => b.count - a.count);

  return {
    totalWords,
    causalCount: totalCausal,
    causalDensity,
    insightCount: totalInsight,
    insightDensity,
    totalCoherenceCount: totalCoherence,
    coherenceRatio,
    causalToInsightRatio,
    depthLevel,
    depthScore,
    detectedCausalWords,
    detectedInsightWords,
    tokenizer: "regex",
    source: "js-fallback",
  };
}

File: nlp/src/coherence_analyzer.py
"""NarrativeRebuild — Coherence Metric Parser (Module 3, Member 1).

A text-analysis engine that scans each writing entry for cause-and-effect
language ("because", "therefore", "since", "consequently", "led to" ...) and
cognitive insight / realization markers ("realize", "understand", "comprehend",
"insight", "learned" ...). It computes causal density, insight density, and an
overall cognitive processing depth ratio, serving as a proxy measure of how
deeply a trauma narrative is being integrated and resolved over time.
"""

import re
from typing import Any, Dict, List

try:
    from nltk.tokenize import word_tokenize
    _TOKENIZER = "nltk"
except ImportError:
    _TOKENIZER = "regex"

# ---------------------------------------------------------------------------
# Causal & Cognitive Insight Lexicons
# ---------------------------------------------------------------------------

CAUSAL_WORDS = {
    "because", "since", "therefore", "thus", "consequently", "hence",
    "cause", "caused", "causes", "causing",
    "reason", "reasons", "why", "wherefore", "inasmuch",
    "result", "results", "resulted", "resulting",
    "effect", "effects", "affect", "affects", "affected", "affecting",
    "leads", "led", "leading",
    "depends", "depended", "depending",
    "outcome", "outcomes", "origin", "originated",
    "driven", "drives", "whereby"
}

# Multi-word causal phrases
CAUSAL_PHRASES = [
    "as a result", "because of", "due to", "in order to", "led to",
    "leading to", "so that", "thanks to", "for this reason"
]

INSIGHT_WORDS = {
    "realize", "realizes", "realized", "realizing", "realization",
    "understand", "understands", "understood", "understanding",
    "comprehend", "comprehends", "comprehended", "comprehending", "comprehension",
    "know", "knows", "knew", "knowing", "known", "knowledge",
    "think", "thinks", "thought", "thinking",
    "conclude", "concludes", "concluded", "concluding", "conclusion",
    "meaning", "meanings", "mean", "means", "meant",
    "insight", "insights", "insightful",
    "clarify", "clarifies", "clarified", "clarifying", "clarity",
    "figure", "figured", "figuring",
    "discern", "discerned", "discerning",
    "learn", "learns", "learned", "learning",
    "discover", "discovers", "discovered", "discovering", "discovery",
    "recognize", "recognizes", "recognized", "recognizing", "recognition",
    "perceive", "perceived", "perceiving", "perception",
    "ponder", "pondered", "pondering",
    "reflect", "reflected", "reflecting", "reflection",
    "acknowledge", "acknowledged", "acknowledging",
    "admit", "admitted", "admitting",
    "accept", "accepted", "accepting", "acceptance",
    "resolve", "resolved", "resolving", "resolution",
    "synthesize", "synthesized", "synthesizing", "synthesis"
}

_WORD_PATTERN = re.compile(r"[a-zA-Z']+")


def _tokenize(text: str) -> List[str]:
    """Tokenize raw text into lowercase word tokens."""
    if _TOKENIZER == "nltk":
        try:
            return [t.lower() for t in word_tokenize(text)]
        except Exception:
            return _WORD_PATTERN.findall(text.lower())
    return _WORD_PATTERN.findall(text.lower())


def analyze_coherence(text: str) -> Dict[str, Any]:
    """Compute cause-and-effect language and cognitive insight density metrics.

    Returns detailed counts, densities (% of total words), coherence depth ratio,
    qualitative classification, and matched keywords for UI highlighting.
    """
    if not text or not text.strip():
        return {
            "totalWords": 0,
            "causalCount": 0,
            "causalDensity": 0.0,
            "insightCount": 0,
            "insightDensity": 0.0,
            "totalCoherenceCount": 0,
            "coherenceRatio": 0.0,
            "causalToInsightRatio": None,
            "depthLevel": "Raw / Descriptive Stance",
            "depthScore": 0,
            "detectedCausalWords": [],
            "detectedInsightWords": [],
            "message": "empty",
            "tokenizer": _TOKENIZER,
        }

    lower_text = text.lower()
    tokens = _tokenize(text)
    total_words = max(len([t for t in tokens if t.isalpha() or "'" in t]), 1)

    # 1. Multi-word phrase matching
    detected_causal: Dict[str, int] = {}
    phrase_causal_count = 0
    for phrase in CAUSAL_PHRASES:
        occurrences = len(re.findall(r"\b" + re.escape(phrase) + r"\b", lower_text))
        if occurrences > 0:
            detected_causal[phrase] = occurrences
            phrase_causal_count += occurrences

    # 2. Token-level matching
    detected_insight: Dict[str, int] = {}
    token_causal_count = 0
    token_insight_count = 0

    for token in tokens:
        if token in CAUSAL_WORDS:
            detected_causal[token] = detected_causal.get(token, 0) + 1
            token_causal_count += 1
        elif token in INSIGHT_WORDS:
            detected_insight[token] = detected_insight.get(token, 0) + 1
            token_insight_count += 1

    total_causal = phrase_causal_count + token_causal_count
    total_insight = token_insight_count
    total_coherence = total_causal + total_insight

    causal_density = round((total_causal / total_words) * 100.0, 2)
    insight_density = round((total_insight / total_words) * 100.0, 2)
    coherence_ratio = round((total_coherence / total_words) * 100.0, 2)

    # Cognitive Processing Depth Level classification
    if coherence_ratio >= 6.0:
        depth_level = "High Cognitive Integration"
        depth_score = min(100, int(coherence_ratio * 12))
    elif coherence_ratio >= 3.0:
        depth_level = "Moderate Cognitive Processing"
        depth_score = int(coherence_ratio * 12)
    elif coherence_ratio > 0.0:
        depth_level = "Emerging Causal Reflection"
        depth_score = max(15, int(coherence_ratio * 12))
    else:
        depth_level = "Raw / Descriptive Stance"
        depth_score = 0

    causal_to_insight = (
        round(total_causal / total_insight, 2) if total_insight > 0 else None
    )

    return {
        "totalWords": total_words,
        "causalCount": total_causal,
        "causalDensity": causal_density,
        "insightCount": total_insight,
        "insightDensity": insight_density,
        "totalCoherenceCount": total_coherence,
        "coherenceRatio": coherence_ratio,
        "causalToInsightRatio": causal_to_insight,
        "depthLevel": depth_level,
        "depthScore": depth_score,
        "detectedCausalWords": [
            {"word": k, "count": v} for k, v in sorted(detected_causal.items(), key=lambda x: -x[1])
        ],
        "detectedInsightWords": [
            {"word": k, "count": v} for k, v in sorted(detected_insight.items(), key=lambda x: -x[1])
        ],
        "tokenizer": _TOKENIZER,
    }
